src/cast_core/src/actors/CrudActor.cpp

```diff
@@ -0,0 +1,1241 @@
+    }
+
+    void run(mongocxx::client_session& session) override {
+        bsoncxx::builder::stream::document filterDocument{};
+        _filterTemplate->view(filterDocument);
+
+        bsoncxx::builder::stream::document replaceDocument{};
+        _replacementTemplate->view(replaceDocument);
+
+        auto filterView = filterDocument.view();
+        auto replaceView = replaceDocument.view();
+        auto ctx = _operation.start();
+        (_onSession) ? _collection.replace_one(session, filterView, replaceView, _options)
+                     : _collection.replace_one(filterView, replaceView, _options);
+        ctx.success();
+    }
+
+private:
+    bool _onSession;
+    mongocxx::collection _collection;
+    std::unique_ptr<DocGenerator> _filterTemplate;
+    std::unique_ptr<DocGenerator> _replacementTemplate;
+    metrics::Operation _operation;
+    mongocxx::options::replace _options;
+};
+
+WriteOpCallback createInsertOne =
+    [](YAML::Node opNode,
+       bool onSession,
+       mongocxx::collection collection,
+       genny::DefaultRandom& rng,
+       metrics::Operation operation) -> std::unique_ptr<WriteOperation> {
+    return std::make_unique<InsertOneOperation>(opNode, onSession, collection, rng, operation);
+};
+
+WriteOpCallback createUpdateOne =
+    [](YAML::Node opNode,
+       bool onSession,
+       mongocxx::collection collection,
+       genny::DefaultRandom& rng,
+       metrics::Operation operation) -> std::unique_ptr<WriteOperation> {
+    return std::make_unique<UpdateOneOperation>(opNode, onSession, collection, rng, operation);
+};
+
+WriteOpCallback createDeleteOne =
+    [](YAML::Node opNode,
+       bool onSession,
+       mongocxx::collection collection,
+       genny::DefaultRandom& rng,
+       metrics::Operation operation) -> std::unique_ptr<WriteOperation> {
+    return std::make_unique<DeleteOneOperation>(opNode, onSession, collection, rng, operation);
+};
+
+WriteOpCallback createDeleteMany =
+    [](YAML::Node opNode,
+       bool onSession,
+       mongocxx::collection collection,
+       genny::DefaultRandom& rng,
+       metrics::Operation operation) -> std::unique_ptr<WriteOperation> {
+    return std::make_unique<DeleteManyOperation>(opNode, onSession, collection, rng, operation);
+};
+
+WriteOpCallback createReplaceOne =
+    [](YAML::Node opNode,
+       bool onSession,
+       mongocxx::collection collection,
+       genny::DefaultRandom& rng,
+       metrics::Operation operation) -> std::unique_ptr<WriteOperation> {
+    return std::make_unique<ReplaceOneOperation>(opNode, onSession, collection, rng, operation);
+};
+
+WriteOpCallback createUpdateMany =
+    [](YAML::Node opNode,
+       bool onSession,
+       mongocxx::collection collection,
+       genny::DefaultRandom& rng,
+       metrics::Operation operation) -> std::unique_ptr<WriteOperation> {
+    return std::make_unique<UpdateManyOperation>(opNode, onSession, collection, rng, operation);
+};
```

Comment:
I *think* you can template this guy to remove all the callback boilerplate:

```c++
template<class P, class C, class O>
C baseCallback =
        [](YAML::Node opNode,
           bool onSession,
           mongocxx::collection collection,
           genny::DefaultRandom& rng,
           metrics::Operation operation) -> std::unique_ptr<P> {
            return std::make_unique<O>(opNode, onSession, collection, rng, operation);
        };
```
and you can move these inline into the ctors map

```c++
bulkWriteConstructors = {
   ...
  {"updateOne", baseCallback<WriteOperation, WriteOpCallback, UpdateOneOperation>},
  ...
};
```

Reply:
Ditto for `OpCallback`s below